### ISSUE TITLE
Question: example on using redux-logic inside the react-boilerplate?

### ISSUE THREAD

**Issue by andreawyss** [jeffbski/redux-logic#20]
Is there and example showing how to use redux-logic inside the react-boilerplate?
The issue is that you cannot have a single rootLogic that imports all the application logic because this will prevent code splitting by route that is built into this boilerplate.

I need some help on how to replace the redux-saga with redux-logic and still maintain the code splitting intact.
Thank you

**Comment by jeffbski** [comment]
I built in features that help with code splitting in the API so using `mergeNewLogic([mergeLogic])` will let you dynamically add in code split logic bundles. It even tracks the refs to the ones you have so if you happen to merge them again it will ignore the ones that you already have and only add the new ones. 

So it should work great once you know how to plug it in. I can take a look at react-boilerplate and see if I can help you figure out what needs to change. Like you said, it is probably a matter of looking for where the sagas are being merged and then you can merge the logic there. 

Are you just using the master version of react-boilerplate? If so I can take a look and figure out what would need to change. I wouldn't think it will take that much.

**Comment by jeffbski** [comment]
I also posted a question in the gitter chat to see if those that are familiar with the boilerplate could point us in the right directions. I can start by searching for saga in the file structure and that should turn up most of what we might need to change.

**Comment by jeffbski** [comment]
Most likely most of the work will be:
 - in the store.js where we will createLogicMiddleware instead of createSagaMiddleware
 - instead of using injectSagas we will be injecting logic using mergeNewLogic
 - we won't need to do anything instead of runSaga, you don't have to run logic they are running once they are merged in.

I'll see if I can create a fork of react-boilerplate that uses redux-logic

It looks like doing `grep -i saga -r app server internals` turns up where sagas are being referenced and so just a few of those files will need to be adapted to redux-logic.

**Comment by jeffbski** [comment]
And to answer your original question you can start out with an empty array of logic if you are loading them all from split bundles. 

I usually add the logicMiddleware instance (from createLogicMiddleware) as a property on store so I can access it later like `store.logicMiddleware`

Then when we want to inject new logic, we simply call store.logicMiddleware.mergeNewLogic([newLogic]); // it will even figure out if a particular logic was already loaded and filter those out for you making it easy.

And that's it, the logic is loaded and ready to use. 

I'll take closer look at creating a fork of react-boilerplate tomorrow.

**Comment by andreawyss** [comment]
Awesome. Thank you for helping out!

**Comment by jeffbski** [comment]
It looks like this gives us a good idea what would need to change https://github.com/mxstbr/react-boilerplate/blob/master/docs/js/remove.md#removing-redux-saga

**Comment by jeffbski** [comment]
@andreawyss I have forked react-boilerplate and have a working version using redux-logic instead of sagas in this branch: https://github.com/jeffbski/react-boilerplate/tree/redux-logic

I haven't changed the tests and docs yet, so that is next.

I like how much simpler the logic for HomePage looks compared to the original saga version.
 - [logic version](https://github.com/jeffbski/react-boilerplate/blob/redux-logic/app/containers/HomePage/logic.js)
 - [saga version](https://github.com/mxstbr/react-boilerplate/blob/master/app/containers/HomePage/sagas.js)

Here is the [full comparison of code changes thus far](https://github.com/mxstbr/react-boilerplate/compare/master...jeffbski:redux-logic)

**Comment by andreawyss** [comment]
Excellent. I will try it out on Monday. Thank you. 
You should put a link in your doc to this branch so other can use it too. 
react-boilerplate is one of the top 3 most popular boilerplates for React.

**Comment by jeffbski** [comment]
Yeah, once it is ready, I can publish it and link to it. I want to get the
docs updated and all. I got the tests working too. And of course they just
pushed a new version 3.4 so I'll want to convert to that while it is fresh
in my mind.

We need to come up with a name for it. Ideas?

react-boilerplate-logic
react-logic-boilerplate
react-boiling-logic
react-boilerlogic
relogic-boilerplate
reactive-logic-boilerplate
redlogic-boilerplate
react-redlogic-boilerplate
react-redux-logic-boilerplate
Other?

**Comment by andreawyss** [comment]
react-boilerplate-logic should be easier to find when searching the original base one. It is a matter of adding the right keywords so it gets discovered by the search engines.

**Comment by jeffbski** [comment]
Yeah, I was coming to the same conclusion so either:

 - react-boilerplate-logic
 - react-boilerplate-redlogic

is probably good for searching.

**Comment by andreawyss** [comment]
Is it a common practice to refer to redux as red?

**Comment by jeffbski** [comment]
not that I have seen, so probably best to stay with the shorter react-boilerplate-logic

**Comment by andreawyss** [comment]
I was able to apply your changes into my project and everything works! Thank you.
The next step would be for you to take a look at the generation of a new container:
internals/generators/container
index.js, saga.js.hsb saga.test.js.hbs

**Comment by jeffbski** [comment]
excellent. Thanks for mentioning the next step (I'm only lightly familiar with react-boilerplate so it helps), I'll jump on that right after I get the tests done (I'm getting close).

Then after the code is all done we'll update the docs and such.

**Comment by jeffbski** [comment]
I pushed more changes to the branch having updated tests and generators. I haven't tried running them yet but they should be ready to go.

**Comment by kopax** [comment]
Hi @jeffbski , I am testing your react-boilerplate integration of redux-logic and I wonder how I could reuse my saga within the process ?

Also, I have an important need to load data before page change, is this now possible ?

**Comment by andreawyss** [comment]
Only 2 tests fail:
``` 
asyncInjectors
    getAsyncInjectors
      given a store, should return all async injectors
      Error: Expected { reduced: 'soon' } to equal { reduced: 'yup' }
    helpers
      injectAsyncLogic
        given a store, it should provide a function to inject logic
        Error: Expected { reduced: 'soon' } to equal { reduced: 'yup' }
```
This is due to the fact that we need to wait for the 'TRIGGER_TEST' action to be precessed by the reducer before testing the updated value in the store.

**Comment by jeffbski** [comment]
@andreawyss thanks for letting me know. They are passing on my box :-( but I think I know what is going on.

Let's double check your redux-logic version. 

What version of redux-logic do you have? is it v0.11.* or v0.10.*? I had just updated it while working on the boilerplate so you could have an old version??

With v0.11 I took out an unnecessary artificial delay that was in the code that was no longer needed, so if you have 0.11 then it should have the data updated without having to wait any.

We can easily fix this if we changed the test to listen to whenComplete before running
```js
// return promise to our test so it knows it is async
return store.logicMiddleware.whenComplete(() => { 
  // perform our checks here, the state will be updated
});
```

but hopefully if you update to v0.11 it should work as is.

Let me know what you find. If you have v0.11 and it is still an issue I can put the whenComplete back in.

**Comment by andreawyss** [comment]
Updated to redux-logic v0.11.4 and all tests are passing. Thank you.

Maybe in the "internals/generators/container/logic.js.hbs" it would be nice if you include commented code sections for validation, transform and process with processOptions.
The one can easily uncomment the desired section and remove the other ones.

**Comment by jeffbski** [comment]
Excellent. That's a great idea about putting in commented sections like that.

**Comment by andreawyss** [comment]
Also "npm run lint" to make sure that all your changes pass the linter.

**Comment by jeffbski** [comment]
@andreawyss I've updated the logic.js.hbs as you suggested, let me know what you think.

**Comment by jeffbski** [comment]
@andreawyss I have completed the rest of the updates (docs etc) for v3.3.3, let me know how it looks to you. I've renamed it react-boilerplate-logic. So it lives at https://github.com/jeffbski/react-boilerplate-logic (although the old URL should also redirect on github). I also merged the redux-logic branch to master to they are now identical. 

I think this is ready to go unless you find any issues. I think I'll wait on doing the 3.4 conversion till a bit later since I have other fish to fry at the moment.
